rpcs3/Emu/Cell/Modules/cellSearch.cpp

```diff
@@ -42,31 +44,91 @@ void fmt_class_string<CellSearchError>::format(std::string& out, u64 arg)
 	});
 }
 
+enum class SearchState
+{
+	NOT_INITIALIZED  = 0,
+	IDLE,
+	IN_PROGRESS,
+	INITIALIZING,
+	CANCELING,
+	FINALIZING,
+};
+
 struct search_info
 {
 	vm::ptr<CellSearchSystemCallback> func;
-	vm::ptr<void> userData;
+	vm::ptr<void> user_data;
+
+	std::unordered_map<std::string, std::string> content_links;
+	SearchState state = SearchState::NOT_INITIALIZED;
+};
+
+struct search_content_t
+{
+	CellSearchContentType type = CELL_SEARCH_CONTENTTYPE_NONE;
+	CellSearchTimeInfo timeInfo;
+	CellSearchContentInfoPath infoPath;
+	union
+	{
+		CellSearchMusicInfo music;
```

Comment:
I strongly against using any kind of union/shared buffer here. aside from the fact that it's UB accessing different members than intended bugs are super annoying to debug due to unexpected behaviours. I prefer unwrapping all members into single default initialized ones at the cost of little extra memory used.
If this was a struct used in vm memory than sure use a union, but if we can avoid it altogether it's better.

Reply:
This union is a textbook example of how to use unions, where the "type" of the item stored is determined via external factors and remains unchanged during the entire lifetime of that specific object. I personally think it's a very fit use case, since it's possible to `memcpy` the entire stuff to vm memory from our memory - just like the original `cellSearch` PR did, albeit there a byte buffer was used instead.

Reply:
ok then.

Reply:
Besides, notice that this is a "fire-and-forget" structure. It's filled once and then HLE does nothing but `memcpy` to vm memory. I'd agree if HLE kept manipulating it, but for copying I wouldn't bother.